### PR TITLE
Support Multiple classification.

### PR DESCRIPTION
**Pull request by XiaoningJin**

### CODE REVIEW

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 1]
```diff
@@ -1,7 +1,115 @@
...
+        self.label_name = key
+        self.label_value_pairs = item.get(key)
+        self.validate()
+     
+    def validate(self):
+        if not self.label_name or not isinstance(self.label_name, str):
+            raise ValueError(f"label name must be type str and value must not empty. {self.label_name}")
+    
+        if not self.label_value_pairs or not isinstance(self.label_value_pairs, dict):
+            raise ValueError(f"label scores must be dict. {self.label_value_pairs}")
```
Comment (DavidZeLiang) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:41]: grammar, casing, format

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 2]
```diff
@@ -1,7 +1,115 @@
...
+        
+        for k, v in self.label_value_pairs.items():
+            if not k or not isinstance(k, str):
+                raise ValueError(f"key of label value paris should be str and not empty: {k}")
+            if not v or not isinstance(v, (float, int)):
```
Comment (DavidZeLiang) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:46]: type

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 3]
```diff
@@ -1,7 +1,115 @@
...
+                raise ValueError(f"value of label value pairs should be float or int.")
+            
+        value_sum = self.value_sum()
```
Comment (DavidZeLiang) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:49]: remove

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 4]
```diff
@@ -1,7 +1,115 @@
...
+       
+        value_key_pairs_number = self.get_key_number_of_value_pairs()
+        if 0 >= value_key_pairs_number:
+            raise ValueError(f"there must be at least one label in value pairs: {value_key_pairs_number}")
+        
+    def value_sum(self):
+        value_sum = sum(v for v in self.label_value_pairs.values())
+        return value_sum
+    
+    def get_value_pairs_keys(self):
```
Comment (DavidZeLiang) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:61]: remove

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 5]
```diff
@@ -1,7 +1,115 @@
...
+            raise ValueError(f"confidence scores must be dict and may not empty: {self.confidence_scores}")
+        
+        for k, v in self.confidence_scores.items():
+            if not k or not isinstance(k, str):
+                raise ValueError(f"confidence score key name must be str and not empty: {k}")
+            if not v or not isinstance(v, (float, int)):
+                raise ValueError(f"confidence score value must be float or int and not empty: {v}")
+            
+        key_number_confidence_scores = self.get_key_number_of_confidence_scores()
+        if 0 >= key_number_confidence_scores:
```
Comment (DavidZeLiang) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:103]: if xxx <= 0:

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 6]
```diff
@@ -1,7 +1,115 @@
...
+            raise ValueError(f"there must be at least one key in confidence score: {self.confidence_scores}")
+        
+    def get_confidence_scores_keys(self):
+        return self.confidence_scores.keys()
+    
+    def get_key_number_of_confidence_scores(self):
```
Comment (DavidZeLiang) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:109]: def __item_count__()

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 7]
```diff
@@ -71,19 +73,126 @@ def images(self, images: str) -> None:
 
 
 class InferenceInput(MopInferenceInput):
-    '''this is a subclass to be compatible with old version'''
+    """this is a subclass to be compatible with old version"""
```
Comment (DavidZeLiang) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:76]: deprecated notation

Comment (XiaoningJin) [reply packages/mop_utils/mop_utils/base_model_wrapper.py]: Done, thanks!

File: packages/mop_utils/mop_utils/base_model_wrapper.py (hunk 6 above)
Comment (XiaoningJin) [reply packages/mop_utils/mop_utils/base_model_wrapper.py]: Directly use "0 == len(self.scores.keys())" instead.

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 8]
```diff
@@ -71,19 +72,66 @@ def images(self, images: str) -> None:
...
+        self.predicted_labels = raw_data_dict.get("predicted_labels", None)
+        self.confidence_scores = raw_data_dict.get("confidence_scores", None)
+        if not self.__predicted_labels__ or not self.__confidence_scores__:
+            raise ValueError(f"MopInferenceOutput: invalid raw data dict: Current value: {raw_data_dict}")
+
+        MopInferenceOutputValidator(self).validate()
+
+    def from_dict(self, raw_data_dict: dict) -> Any:
+        """
+        Deprecated: This version will be deprecated.
```
Comment (DavidZeLiang) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:128]: This method

File: packages/mop_utils/mop_utils/base_model_wrapper.py [hunk 9]
```diff
@@ -71,19 +72,66 @@ def images(self, images: str) -> None:
 
 
 class InferenceInput(MopInferenceInput):
```
Comment (DavidZeLiang) [inline packages/mop_utils/mop_utils/base_model_wrapper.py:74]: delete

File: sample/src/inference.py [hunk 10]
```diff
@@ -1,12 +1,14 @@
...
 from typing import List, Dict
 
 import nltk
-from mop_utils.base_model_wrapper import BaseModelWrapper, InferenceInput, InferenceOutput
+from mop_utils.base_model_wrapper import BaseModelWrapper, MopInferenceInput, MopInferenceOutput
+from mop_utils.util import TextAnalysisInput, ImageAnalysisInput, AcsTextResponse, AcsImageResponse
 
 
-# from base_model_wrapper import  BaseModelWrapper, InferenceInput, InferenceOutput
+# from base_model_wrapper import  BaseModelWrapper, MopInferenceInput, MopInferenceOutput
```
Comment (DavidZeLiang) [inline sample/src/inference.py:11]: delete

File: sample/src/inference.py [hunk 11]
```diff
@@ -33,16 +35,182 @@ def inference_batch(self, items: List[Dict]) -> List[Dict]:
...
 
-    def convert_mop_input_to_model_input(self, mop_input: InferenceInput, **kwargs) -> Dict:
+    def convert_mop_input_to_model_input(self, mop_input: MopInferenceInput, **kwargs) -> Dict:
         return {"data": mop_input.text}
 
-    def convert_model_output_to_mop_output(self, customized_output: Dict, **kwargs) -> InferenceOutput:
-        inference_output = InferenceOutput()
+    def convert_model_output_to_mop_output(self, customized_output: Dict, **kwargs) -> MopInferenceOutput:
         customized_output = float(customized_output.get('score'))
         print(f'customized_output = {customized_output}')
```
Comment (DavidZeLiang) [inline sample/src/inference.py:43]: refine

File: sample/src/inference.py [hunk 12]
```diff
@@ -33,16 +35,182 @@ def inference_batch(self, items: List[Dict]) -> List[Dict]:
         scores = [{'score': x[1]} for x in predicts]
         return scores
 
-    def convert_mop_input_to_model_input(self, mop_input: InferenceInput, **kwargs) -> Dict:
+    def convert_mop_input_to_model_input(self, mop_input: MopInferenceInput, **kwargs) -> Dict:
```
Comment (DavidZeLiang) [inline sample/src/inference.py:38]: add a description in inference method

File: sample/src/inference.py [hunk 13]
```diff
@@ -1,12 +1,14 @@
 import json
```
Comment (DavidZeLiang) [inline sample/src/inference.py:1]: /sample/sample-model/binary-text-model/src
/sample/sample-model/binary-text-model/model
/sample/sample-model/binary-text-model/readme.md

/sample/sample-dataset/Image/data
/sample/sample-dataset/Text/data
/sample/sample-dataset/ImageAndText/data

File: sample/src/inference.py [hunk 14]
```diff
@@ -33,16 +35,182 @@ def inference_batch(self, items: List[Dict]) -> List[Dict]:
...
         return scores
 
-    def convert_mop_input_to_model_input(self, mop_input: InferenceInput, **kwargs) -> Dict:
+    def convert_mop_input_to_model_input(self, mop_input: MopInferenceInput, **kwargs) -> Dict:
         return {"data": mop_input.text}
 
-    def convert_model_output_to_mop_output(self, customized_output: Dict, **kwargs) -> InferenceOutput:
-        inference_output = InferenceOutput()
+    def convert_model_output_to_mop_output(self, customized_output: Dict, **kwargs) -> MopInferenceOutput:
         customized_output = float(customized_output.get('score'))
```
Comment (DavidZeLiang) [inline sample/src/inference.py:42]: probability = float(customized_output.get('score'))

File: sample/src/inference.py [hunk 15]
```diff
@@ -33,16 +35,182 @@ def inference_batch(self, items: List[Dict]) -> List[Dict]:
...
+        d = {
+            "confidence_scores": {
+                "identity_hate": {
+                    "positive": customized_output,
+                    "negative": 1 - customized_output
+                }
+            },
+            "predicted_labels": {
+                "identity_hate": {
+                    "positive": 1,
```
Comment (DavidZeLiang) [inline sample/src/inference.py:53]: "positive": 1 if probability > 0.5 else 0,
"negative" 1 - (1 if probability > 0.5 else 0)

File: packages/mop_utils/mop_utils/util.py [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+
+
+class AcsImageResponse:
+    hate = AnalysisResult()
+    self_harm = AnalysisResult()
+    sexual = AnalysisResult()
+    violence = AnalysisResult()
+
+
+class AcsTextResponse:
```
Comment (DavidZeLiang) [inline packages/mop_utils/mop_utils/util.py:26]: add description to illustrate all four fields are required.

File: sample/src/inference.py [hunk 17]
```diff
@@ -33,16 +35,182 @@ def inference_batch(self, items: List[Dict]) -> List[Dict]:
...
+        But it is mandatory if you want to release the model to AACS service.
+        @param req: ACS text request
+        @type req: TextAnalysisInput
+        @return: Model inference input
+        @rtype: object
+        """
+        return {"data": req.text}
+
+    def convert_model_inference_output_to_acs_text_response(self, out: object) -> AcsTextResponse:
+        """
```
Comment (DavidZeLiang) [inline sample/src/inference.py:73]: out: {'score': score}
# for hate:
score = out["score"]
analysis_result = AnalysisResult()
analysis_result.harmful_score = score
analysis_result.severity_level = floor(score * 8) # 0-7

res = AcsTextResponse()
res.hate = analysis_result
res.self-harm = analysis_result
..
retrurn res

File: sample/src/inference.py [hunk 18]
```diff
@@ -33,16 +35,182 @@ def inference_batch(self, items: List[Dict]) -> List[Dict]:
...
+        acs_text_resp.self_harm.severity_level = out["self-harm"]["class"]
+        
+        acs_text_resp.hate.harmful_score = out["hate"]["scores"][0]
+        acs_text_resp.hate.severity_level = out["hate"]["class"]
+        
+        acs_text_resp.sexual.harmful_score = out["sexual"]["scores"][0]
+        acs_text_resp.sexual.severity_level = out["sexual"]["class"]
+        return acs_text_resp
+
+    def convert_acs_image_request_to_model_inference_input(self, req: ImageAnalysisInput) -> object:
```
Comment (DavidZeLiang) [inline sample/src/inference.py:137]: remove since this model is for text, not for image

File: sample/src/inference.py [hunk 19]
```diff
@@ -52,10 +220,77 @@ def convert_model_output_to_mop_output(self, customized_output: Dict, **kwargs)
     c_output = model_wrapper.inference(customized_input)
```
Comment (DavidZeLiang) [inline sample/src/inference.py:220]: reorganize the output print

**Review by DavidZeLiang — COMMENTED** [review]
ok for release

**Review by DavidZeLiang — APPROVED** [review]
ok for release

**Pull request merged**